homeassistant/components/imap_email_content/sensor.py

```diff
@@ -254,22 +267,30 @@ def get_msg_text(email_message):
     def update(self) -> None:
         """Read emails and publish state change."""
         email_message = self._email_reader.read_next()
-
-        if email_message is None:
-            self._message = None
-            self._state_attributes = {}
-            return
-
-        if self.sender_allowed(email_message):
-            message = EmailContentSensor.get_msg_subject(email_message)
-
-            if self._value_template is not None:
-                message = self.render_template(email_message)
-
-            self._message = message
-            self._state_attributes = {
-                ATTR_FROM: EmailContentSensor.get_msg_sender(email_message),
-                ATTR_SUBJECT: EmailContentSensor.get_msg_subject(email_message),
-                ATTR_DATE: email_message["Date"],
-                ATTR_BODY: EmailContentSensor.get_msg_text(email_message),
-            }
+        while (
+            self._last_id is None or self._last_id != self._email_reader.last_unread_id
+        ):
+            if email_message is None:
+                self._message = None
+                self._state_attributes = {}
+                return
+
+            self._last_id = self._email_reader.last_id
+
+            if self.sender_allowed(email_message):
+                message = EmailContentSensor.get_msg_subject(email_message)
+
+                if self._value_template is not None:
+                    message = self.render_template(email_message)
+
+                self._message = message
+                self._state_attributes = {
+                    ATTR_FROM: EmailContentSensor.get_msg_sender(email_message),
+                    ATTR_SUBJECT: EmailContentSensor.get_msg_subject(email_message),
+                    ATTR_DATE: email_message["Date"],
+                    ATTR_BODY: EmailContentSensor.get_msg_text(email_message),
+                }
+
```

Comment:
It seems like this should only happen if we are going to break out of the loop instead of processing it and than throwing it away when we hit the next one?

Reply:
We do not know if an unseen mail has an allowed sender unless we process it. The current code base uses a custom `EmailReader` class as backend.
The code in `dev` needs several update iterations before it reaches the last one.
The improvement here is to iterate the messages until we reach the last one in a single update cycle.